# COVID-19 Search Trends & Hospitalizations


So we need to do a mixture of code files and notebooks cause they want to run some of our stuff easily. 
I'm guessing we can simply put our good stuff here and play around in the code files.


In [12]:
import numpy as np
import pandas as pd

First, load datasets directly into 2D numpy arrays. Two different seach data arrays were used: one with string values for the feature names and the other with float numbers for preprocessing


In [13]:
search_data = np.genfromtxt('2020_US_weekly_symptoms_dataset.csv', delimiter=',')
search_data_info = np.genfromtxt('2020_US_weekly_symptoms_dataset.csv', delimiter=',', dtype=None, encoding='utf-8')
hospitalization_data = np.genfromtxt('aggregated_cc_by.csv', delimiter=',', dtype=None, encoding='utf-8')

search_data_no_head = search_data[1:, :]
hospitalization_data_no_head = hospitalization_data[1:, :]
search_data_name = search_data_info[:, 0:8]
search_data_symptoms = search_data_info[0, 9:]

**Cleaning up the datasets**: Any symptoms with less than 50% of their data filled were removed. Subregion 2 columns also are removed since they didn't categorize any smaller subregions. 


In [14]:
search_data = np.delete(search_data_no_head, np.count_nonzero(np.isnan(search_data_no_head), axis=0) >
                        0.5 * np.shape(search_data_no_head)[0], axis=1)
search_data_symptoms = search_data_symptoms[~(np.count_nonzero(np.isnan(search_data_no_head[:, 9:]), axis=0) >
                                              0.5 * np.shape(search_data_no_head)[0])]
search_data_name = np.delete(search_data_name, [5, 6], axis=1)

**Normalizing the Data:** Since the dataset was normalized to each specific region, we desire to re-normalize it again to be comparable across regions. To do this we subtracted the median value of each symptom seperated by region from each data point or de-median'd the data. There were 16 different state regions with 38 week

In [15]:
search_data[np.isnan(search_data)] = 0

for i in range(0, 16):
    lower = i*38
    upper = lower+38
    median = np.median(search_data[lower:upper, :], axis=0)# MEDIAN ARRAY
    for j in range(search_data.shape[1]):
        search_data[lower:upper, j] = search_data[lower:upper, j] - median[j]

search_data_with_head = np.concatenate(([search_data_symptoms], search_data), axis=0)

clean_search_data = np.concatenate((search_data_name, search_data_with_head), axis=1)

**Cleaning up the datasets**: Hospitalization data contained entries for several countries and American states. Only states included in the search symptoms dataset were taken out and any states with very little hospitalization data that was not reflective of their COVID-19 case numbers were removed as well. This included District of Columbia, Delaware and West Virginia.

In [6]:
hospital_columns = [0, 1, 2, 16, 17]
american_states = ['AK', 'HI', 'ID', 'ME', 'MT', 'ND', 'NE', 'NH', 'NM', 'RI', 'SD', 'VT', 'WY']

hospitalization_data = np.take(hospitalization_data, hospital_columns, axis=1)

us_hospitalization_data = np.empty((1, 5))

for state in american_states:
    covid_region_code = f"US-{state}"
    state_data = np.delete(hospitalization_data, ~(np.any(hospitalization_data == covid_region_code, axis=1)), axis=0)
    us_hospitalization_data = np.append(us_hospitalization_data, state_data, axis=0)

us_hospitalization_data = us_hospitalization_data[1:,:]
print(us_hospitalization_data)

[['US-AK' 'Alaska' '2020-03-06' '0.0' '0.0']
 ['US-AK' 'Alaska' '2020-03-07' '0.0' '0.0']
 ['US-AK' 'Alaska' '2020-03-08' '0.0' '0.0']
 ...
 ['US-WY' 'Wyoming' '2020-10-04' '1.0' '282.0']
 ['US-WY' 'Wyoming' '2020-10-05' '12.0' '294.0']
 ['US-WY' 'Wyoming' '2020-10-06' '10.0' '304.0']]


**Cleaning up the datasets**: Remove states and dates with no hospitalization data in the search data.

In [7]:
clean_search_data = np.delete(clean_search_data, np.any(clean_search_data == "US-WV", axis=1), axis=0)
clean_search_data = np.delete(clean_search_data, np.any(clean_search_data == "US-DC", axis=1), axis=0)
clean_search_data = np.delete(clean_search_data, np.any(clean_search_data == "US-DE", axis=1), axis=0)

remove_dates = ['01-06', '01-13', '01-20', '01-27', '02-03', '02-10', '02-17', '02-24']
for date in remove_dates:
    full_date = f"2020-{date}"
    clean_search_data = np.delete(clean_search_data, np.any(clean_search_data == full_date, axis=1), axis=0)

print(clean_search_data)

[['open_covid_region_code' 'country_region_code' 'country_region' ...
  'symptom:Urinary urgency' 'symptom:Ventricular fibrillation'
  'symptom:Viral pneumonia']
 ['US-AK' 'US' 'United States' ... '-6.055' '3.410000000000001'
  '7.544999999999999']
 ['US-AK' 'US' 'United States' ... '-6.055' '-7.05' '12.275000000000002']
 ...
 ['US-WY' 'US' 'United States' ... '-6.055' '-7.05' '-6.575']
 ['US-WY' 'US' 'United States' ... '-3.3749999999999996' '-7.05' '-6.575']
 ['US-WY' 'US' 'United States' ... '-6.055' '-3.82' '-6.575']]


**Merging the datasets**: The hospitalization data is given daily hence need be converted to weekly. Daily data was summed together and aggregated to the Monday of that week starting from March-2-2020 to September-21-2020. 

In [8]:
us_hospitalization_data = pd.DataFrame(us_hospitalization_data, columns=['open_covid_region_code', 'sub_region_1',
                                                                         'date', 'hospitalized_new',
                                                                         'hospitalized_cumulative'])

us_hospitalization_data['date'] = us_hospitalization_data['date'].astype('datetime64[ns]')
us_hospitalization_data['hospitalized_new'] = us_hospitalization_data['hospitalized_new'].astype('float')
us_hospitalization_data['hospitalized_cumulative'] = us_hospitalization_data['hospitalized_cumulative'].astype('float')

weekly_hospitalization_data = us_hospitalization_data.groupby(by=["open_covid_region_code", 'sub_region_1']).resample(
    "W-MON",
    label='left',
    closed='left',
    on='date').sum().reset_index()

weekly_hospitalization_data

open_covid_region_code sub_region_1       date  hospitalized_new  \
0                    US-AK       Alaska 2020-03-02               0.0   
1                    US-AK       Alaska 2020-03-09               0.0   
2                    US-AK       Alaska 2020-03-16               0.0   
3                    US-AK       Alaska 2020-03-23               6.0   
4                    US-AK       Alaska 2020-03-30              14.0   
..                     ...          ...        ...               ...   
412                  US-WY      Wyoming 2020-09-07               9.0   
413                  US-WY      Wyoming 2020-09-14              15.0   
414                  US-WY      Wyoming 2020-09-21              19.0   
415                  US-WY      Wyoming 2020-09-28              20.0   
416                  US-WY      Wyoming 2020-10-05              22.0   

     hospitalized_cumulative  
0                        0.0  
1                        0.0  
2                        0.0  
3                       18.0  
4                       83.0  
..                       ...  
412                   1625.0  
413                   1656.0  
414                   1784.0  
415                   1920.0  
416                    598.0  

[417 rows x 5 columns]

**Cleaning the datasets:** Extra weeks were included in the hospitalization data not in the search data hence they were excluded. 

In [9]:
weekly_hospitalization_data = weekly_hospitalization_data.drop(weekly_hospitalization_data[(weekly_hospitalization_data.
                                                                                            date == '2020-09-28') |
                                                                                           (weekly_hospitalization_data.
                                                                                            date == '2020-10-05') |
                                                                                           (weekly_hospitalization_data.
                                                                                            date == '2020-02-24')].
                                                               index).reset_index(drop=True)
weekly_hospitalization_data

open_covid_region_code sub_region_1       date  hospitalized_new  \
0                    US-AK       Alaska 2020-03-02               0.0   
1                    US-AK       Alaska 2020-03-09               0.0   
2                    US-AK       Alaska 2020-03-16               0.0   
3                    US-AK       Alaska 2020-03-23               6.0   
4                    US-AK       Alaska 2020-03-30              14.0   
..                     ...          ...        ...               ...   
385                  US-WY      Wyoming 2020-08-24               8.0   
386                  US-WY      Wyoming 2020-08-31               4.0   
387                  US-WY      Wyoming 2020-09-07               9.0   
388                  US-WY      Wyoming 2020-09-14              15.0   
389                  US-WY      Wyoming 2020-09-21              19.0   

     hospitalized_cumulative  
0                        0.0  
1                        0.0  
2                        0.0  
3                       18.0  
4                       83.0  
..                       ...  
385                   1497.0  
386                   1543.0  
387                   1625.0  
388                   1656.0  
389                   1784.0  

[390 rows x 5 columns]

**Merging the datasets:** Now that both datasets have 390 rows aligned by state and week, we can take the last two columns of the hospitalzation data and concatenate/append it to the search data. 

In [10]:
search_dataframe = pd.DataFrame(clean_search_data[1:, :], columns=clean_search_data[0, :])

final_covid_dataframe = pd.concat([search_dataframe, weekly_hospitalization_data[['hospitalized_new',
                                                                                  'hospitalized_cumulative']]],
                                  axis=1)

final_covid_dataset = np.append(clean_search_data[1:, :], np.array(weekly_hospitalization_data
                                                                   [['hospitalized_new',
                                                                     'hospitalized_cumulative']]), axis=1)
final_covid_dataset
final_covid_dataframe

open_covid_region_code country_region_code country_region sub_region_1  \
0                    US-AK                  US  United States       Alaska   
1                    US-AK                  US  United States       Alaska   
2                    US-AK                  US  United States       Alaska   
3                    US-AK                  US  United States       Alaska   
4                    US-AK                  US  United States       Alaska   
..                     ...                 ...            ...          ...   
385                  US-WY                  US  United States      Wyoming   
386                  US-WY                  US  United States      Wyoming   
387                  US-WY                  US  United States      Wyoming   
388                  US-WY                  US  United States      Wyoming   
389                  US-WY                  US  United States      Wyoming   

    sub_region_1_code        date symptom:Allergic conjunctivitis  \
0               US-AK  2020-03-02                          -3.375   
1               US-AK  2020-03-09                           8.235   
2               US-AK  2020-03-16              3.0149999999999997   
3               US-AK  2020-03-23                          -3.375   
4               US-AK  2020-03-30              3.7350000000000003   
..                ...         ...                             ...   
385             US-WY  2020-08-24             0.04499999999999993   
386             US-WY  2020-08-31                          -3.375   
387             US-WY  2020-09-07             -1.1749999999999998   
388             US-WY  2020-09-14                          -3.375   
389             US-WY  2020-09-21                          -3.375   

    symptom:Angular cheilitis symptom:Aphonia symptom:Auditory hallucination  \
0                       -6.43          -9.035              4.450000000000001   
1                       -6.43          -1.125                          -3.85   
2                       -6.43          -1.085                          -3.85   
3                       -6.43          -9.035                            2.6   
4                       -6.43          -9.035                           2.69   
..                        ...             ...                            ...   
385                     -6.43          -9.035             0.6800000000000002   
386                     -6.43          -9.035                          -3.85   
387                     -4.18          -9.035            -1.4300000000000002   
388                     -6.43          -9.035            -0.2400000000000002   
389                     -6.43          -9.035            -1.1099999999999999   

     ... symptom:Pulmonary edema  symptom:Rectal pain   symptom:Rumination  \
0    ...       5.335000000000001   1.6400000000000006  0.27000000000000046   
1    ...      3.6050000000000004   0.5700000000000003                 1.87   
2    ...       5.074999999999999   0.8300000000000001                 5.14   
3    ...       4.245000000000001  -0.1900000000000004   4.7299999999999995   
4    ...       6.415000000000001                 -6.5   2.3899999999999997   
..   ...                     ...                  ...                  ...   
385  ...    -0.04499999999999993                 -6.5                -6.71   
386  ...                  -3.965                -3.87                -3.86   
387  ...     -1.4849999999999999                 -6.5                -3.85   
388  ...                  -3.965                -4.04                -2.83   
389  ...                  -3.965                 -6.5                -4.47   

    symptom:Shallow breathing      symptom:Stridor symptom:Urinary urgency  \
0                        -5.9    7.245000000000001                  -6.055   
1          2.1799999999999997                8.755                  -6.055   
2                        5.01                7.725                  -6.055   
3                        3.3